<a href="https://colab.research.google.com/github/Lucas-Masaba/AI_ML/blob/main/Data_Cleaning_Introduced_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning: Missing Values


> "Garbage in, garbage out."  

Let's clean up the garbage!

In this Notebook we will:

1. Review dropping unnecessary columns and duplicates, correcting datatypes, and fixing poorly formed categories
2. Identify columns with missing values
3. Choose and implement strategies for dealing with missing values.



### Initial Inspection
Phase 1: What data have we been provided?
The stakeholders have provided us with two links:

* [Share-URL to a .csv file](https://drive.google.com/file/d/1Jach7HsZVywhJnUJmkyqje52ho_0VJgo/view?usp=drive_link)  

A spreadsheet of various features of homes in their town, as well as the price of the house at the time of sale.  Add this to your drive or load to the colab session

* [Data Dictionary](https://docs.google.com/document/d/1nmnel7g35aMOl0mKiSsTHXT8wRzbJ1EktKNqYFEmpWE/edit?usp=sharing)
A data dictionary is a document that lists the name and explanation for every feature in a dataset.

**Data Dictionary**  
As Data Scientists, we often work with data that is outside of our domain of expertise. In order for us to quickly get acclimated to the domain of the project/task and to learn the business value of each feature, we may be provided with a data dictionary. Ideally, the data dictionary will clarify any abbreviations or codes that are used in the data. If possible, you could consult the stakeholders for additional clarification if needed. It is a good idea to preview the data dictionary at the onset of a project and to keep it readily available as a reference.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Import Pandas
import pandas as pd

In [ ]:
# Load the Data
df = pd.read_csv('/content/drive/MyDrive/AI_ML_Data_Analytics/slides/Week 3/ames-housing-dojo.csv')
df.head()

Unnamed: 0        PID MS Zoning  Lot Frontage  Lot Area Street Alley  \
0           0  907227090        RL            60      7200   Pave   NaN   
1           1  527108010        RL           134     19378   Pave   NaN   
2           2  534275170        RL            -1     12772   Pave   NaN   
3           3  528104050        RL           114     14803   Pave   NaN   
4           4  533206070        FV            32      3784   Pave  Pave   

  Utilities Neighborhood Bldg Type  ... Garage Type  Garage Yr Blt  \
0    AllPub      CollgCr      1Fam  ...      Detchd         1977.0   
1    AllPub      Gilbert      1Fam  ...      Attchd         2006.0   
2    AllPub        NAmes      1Fam  ...      Attchd         1960.0   
3    AllPub      NridgHt      1Fam  ...      Attchd         2007.0   
4    AllPub      Somerst    TwnhsE  ...      Attchd         2006.0   

   Garage Cars  Garage Area  Garage Qual Garage Cond Paved Drive  Fence  \
0          1.0        297.0           TA          TA           Y  MnPrv   
1          2.0        576.0           TA          TA           Y    NaN   
2          1.0        301.0           TA          TA           Y    NaN   
3          3.0       1220.0           TA          TA           Y    NaN   
4          2.0        476.0           TA          TA           Y    NaN   

   Date Sold SalePrice  
0    03-2006  119900.0  
1    03-2006  320000.0  
2    04-2007  151500.0  
3    06-2008  385000.0  
4    02-2007  193800.0  

[5 rows x 38 columns]

In [ ]:
# How many rows and columns?


In [ ]:
# What are the datatypes?

In [ ]:
# What are the column names?


In [ ]:
# General information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2959 entries, 0 to 2958
Data columns (total 38 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      2959 non-null   int64  
 1   PID             2959 non-null   int64  
 2   MS Zoning       2959 non-null   object 
 3   Lot Frontage    2959 non-null   int64  
 4   Lot Area        2959 non-null   int64  
 5   Street          2959 non-null   object 
 6   Alley           201 non-null    object 
 7   Utilities       2959 non-null   object 
 8   Neighborhood    2959 non-null   object 
 9   Bldg Type       2959 non-null   object 
 10  House Style     2959 non-null   object 
 11  Overall Qual    2959 non-null   int64  
 12  Overall Cond    2959 non-null   int64  
 13  Year Built      2959 non-null   int64  
 14  Year Remod/Add  2959 non-null   int64  
 15  Exter Qual      2959 non-null   object 
 16  Exter Cond      2959 non-null   object 
 17  Bsmt Unf SF     2958 non-null   f

**Changing a data type**

Upon closer inspection, we find that one of our data types is incorrect. The "Gr Liv Area" is defined in the data dictionary as "Above grade (ground) living area square feet." Square footage should be a numeric value, but the data type is "object" (string).

In [ ]:
# Check the datatype for a particular column
df['Gr Liv Area'].dtypes

dtype('O')

In [ ]:
# We can convert using the atype method
df['Gr Liv Area'].astype(float)


ValueError: ignored

We have an error pointing to incompatible datatype conversions. Specifically some strings cannnot be converted to floats.  
Let us inspect the column further

In [ ]:
# Check the value_counts for Living Area Sqft
df['Gr Liv Area'].value_counts()

864sqft     41
1092sqft    27
1040sqft    25
1456sqft    20
1200sqft    18
            ..
2799sqft     1
1778sqft     1
1386sqft     1
2004sqft     1
1789sqft     1
Name: Gr Liv Area, Length: 1292, dtype: int64

We can see that the values have the string **sqft** at the end. What we want is numbers so let us modify the value using the string **replace()** method.

In [ ]:
test_string = "hello world"

In [ ]:
test_string.replace("hello", "hi")

'hi world'

In [ ]:
"864sqft".replace("sqft",'')

'864'

In [ ]:
# Replace Sqft with nothing
df['Gr Liv Area'] = df['Gr Liv Area'].str.replace("sqft",'')
# Preview the first 5 values for the column to verify the change
df['Gr Liv Area'].head()


0     864
1    2462
2     958
3    2084
4    1565
Name: Gr Liv Area, dtype: object

In [ ]:
# let us try to change the column datatype again
# Convert the column to a float
df['Gr Liv Area'] = df['Gr Liv Area'].astype(float)
# Confirm the data type of the column
df['Gr Liv Area'].dtype


dtype('float64')

In [ ]:
df.head(3)

Unnamed: 0        PID MS Zoning  Lot Frontage  Lot Area Street Alley  \
0           0  907227090        RL            60      7200   Pave   NaN   
1           1  527108010        RL           134     19378   Pave   NaN   
2           2  534275170        RL            -1     12772   Pave   NaN   

  Utilities Neighborhood Bldg Type  ... Garage Type  Garage Yr Blt  \
0    AllPub      CollgCr      1Fam  ...      Detchd         1977.0   
1    AllPub      Gilbert      1Fam  ...      Attchd         2006.0   
2    AllPub        NAmes      1Fam  ...      Attchd         1960.0   

   Garage Cars  Garage Area  Garage Qual Garage Cond Paved Drive  Fence  \
0          1.0        297.0           TA          TA           Y  MnPrv   
1          2.0        576.0           TA          TA           Y    NaN   
2          1.0        301.0           TA          TA           Y    NaN   

   Date Sold SalePrice  
0    03-2006  119900.0  
1    03-2006  320000.0  
2    04-2007  151500.0  

[3 rows x 38 columns]

### Deleting columns  
It is also common to remove unreasonable columns from your data. Of course this decision may require domain knowledge to validate your assumptions. For example the column **Unnamed: 0** may not be useful for our analysis.

In [ ]:
# Dropping unnamed 0 (permanently)
df = df.drop(columns=['Unnamed: 0'])
df.head()

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
0  907227090        RL            60      7200   Pave   NaN    AllPub   
1  527108010        RL           134     19378   Pave   NaN    AllPub   
2  534275170        RL            -1     12772   Pave   NaN    AllPub   
3  528104050        RL           114     14803   Pave   NaN    AllPub   
4  533206070        FV            32      3784   Pave  Pave    AllPub   

  Neighborhood Bldg Type House Style  ...  Garage Type  Garage Yr Blt  \
0      CollgCr      1Fam      1Story  ...       Detchd         1977.0   
1      Gilbert      1Fam      2Story  ...       Attchd         2006.0   
2        NAmes      1Fam      1Story  ...       Attchd         1960.0   
3      NridgHt      1Fam      1Story  ...       Attchd         2007.0   
4      Somerst    TwnhsE      1Story  ...       Attchd         2006.0   

   Garage Cars  Garage Area Garage Qual Garage Cond  Paved Drive  Fence  \
0          1.0        297.0          TA          TA            Y  MnPrv   
1          2.0        576.0          TA          TA            Y    NaN   
2          1.0        301.0          TA          TA            Y    NaN   
3          3.0       1220.0          TA          TA            Y    NaN   
4          2.0        476.0          TA          TA            Y    NaN   

  Date Sold  SalePrice  
0   03-2006   119900.0  
1   03-2006   320000.0  
2   04-2007   151500.0  
3   06-2008   385000.0  
4   02-2007   193800.0  

[5 rows x 37 columns]

### Renaming Columns
Renaming columns can make your life easier. Deciding which columns to rename and how to rename them is subjective, but we have selected some very abbreviated column names or ambiguous columns to rename for easier interpretation:

* "Year Remod/Add" is really "Year Remodeled"
* Several Features abbreviate square feet as SF, which we will replace with Sqft:
  *  "Bsmt Unf SF" is really "Bsmt Unfinished Sqft"
  * "Total Bsmt SF" is really "Total Bsmt Sqft"
* "TotRms AbvGrd" is really "Total Rooms"
There are also some misleading/confusing column names:

* "Gr Liv Area" is really the sqft of living area (above ground), which will call "Living Area Sqft"  

While there are other techniques for renaming, we recommend starting by creating a dictionary with the old name matched to the new name.

In [ ]:
# Create a dictionary using old column name : new column name format
rename_dict = {"Year Remod/Add":"Year Remodeled",
               "Bsmt Unf SF": "Bsmt Unf Sqft",
               "Total Bsmt SF": "Total Bsmnt Sqft",
               "TotRms AbvGrd": "Total Rooms",
               "Gr Liv Area":"Living Area Sqft"}
rename_dict

{'Year Remod/Add': 'Year Remodeled',
 'Bsmt Unf SF': 'Bsmt Unf Sqft',
 'Total Bsmt SF': 'Total Bsmnt Sqft',
 'TotRms AbvGrd': 'Total Rooms',
 'Gr Liv Area': 'Living Area Sqft'}

In [ ]:
# dictionary substitution using rename method
df = df.rename(rename_dict,axis=1)
df.head()

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
0  907227090        RL            60      7200   Pave   NaN    AllPub   
1  527108010        RL           134     19378   Pave   NaN    AllPub   
2  534275170        RL            -1     12772   Pave   NaN    AllPub   
3  528104050        RL           114     14803   Pave   NaN    AllPub   
4  533206070        FV            32      3784   Pave  Pave    AllPub   

  Neighborhood Bldg Type House Style  ...  Garage Type  Garage Yr Blt  \
0      CollgCr      1Fam      1Story  ...       Detchd         1977.0   
1      Gilbert      1Fam      2Story  ...       Attchd         2006.0   
2        NAmes      1Fam      1Story  ...       Attchd         1960.0   
3      NridgHt      1Fam      1Story  ...       Attchd         2007.0   
4      Somerst    TwnhsE      1Story  ...       Attchd         2006.0   

   Garage Cars  Garage Area Garage Qual Garage Cond  Paved Drive  Fence  \
0          1.0        297.0          TA          TA            Y  MnPrv   
1          2.0        576.0          TA          TA            Y    NaN   
2          1.0        301.0          TA          TA            Y    NaN   
3          3.0       1220.0          TA          TA            Y    NaN   
4          2.0        476.0          TA          TA            Y    NaN   

  Date Sold  SalePrice  
0   03-2006   119900.0  
1   03-2006   320000.0  
2   04-2007   151500.0  
3   06-2008   385000.0  
4   02-2007   193800.0  

[5 rows x 37 columns]

In [ ]:
# check if columns are renamed

# Remove Duplicates

We don't want any duplicate entries in our data.  This will skew our analysis and confuse our predictive models.  Duplicate entries are quite common in data that has not been cleaned.

We can use `df.duplicated()` to show whether rows are duplicates.

In [ ]:
# Check for duplicates
duplicated_rows  = df.duplicated()
duplicated_rows

0       False
1       False
2       False
3       False
4       False
        ...  
2954    False
2955    False
2956    False
2957    False
2958    False
Length: 2959, dtype: bool

Well, we don't want to have to look row by row, so let's use `.sum()` to add up all of the `True` values.  When using `.sum()`, a `True` value will evaluate to a 1.

In [ ]:
# Count the duplicates
df.duplicated().sum()
# OR
# duplicated_rows.sum()

7

We have 7 duplicated rows.  We can simply remove those with df.drop_duplicates().

In [ ]:
# Visually checking the duplicate rows
df[duplicated_rows]

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
869   535153150        RL            76      9120   Pave   NaN    AllPub   
1019  921205030        RL            88     11443   Pave   NaN    AllPub   
1867  908103280        RL            65      6500   Pave   NaN    AllPub   
2029  526351010        RL            81     14267   Pave   NaN    AllPub   
2203  923230040        RL            63      9297   Pave   NaN    AllPub   
2306  907262070        RL            72      7226   Pave   NaN    AllPub   
2552  528174020        RL            34      3901   Pave   NaN    AllPub   

     Neighborhood Bldg Type House Style  ...  Garage Type  Garage Yr Blt  \
869         NAmes      1Fam      1Story  ...       Attchd         1958.0   
1019       Timber      1Fam      1Story  ...       Attchd         2005.0   
1867      Edwards      1Fam      1Story  ...       Detchd         1991.0   
2029        NAmes      1Fam      1Story  ...       Attchd         1958.0   
2203      Mitchel    Duplex      1Story  ...       Detchd         1976.0   
2306      CollgCr      1Fam      2Story  ...       Attchd         2003.0   
2552      NridgHt     Twnhs      1Story  ...       Attchd         2005.0   

      Garage Cars  Garage Area Garage Qual Garage Cond  Paved Drive  Fence  \
869           2.0        433.0          TA          TA            Y    NaN   
1019          3.0        880.0          TA          TA            Y    NaN   
1867          2.0        480.0          TA          TA            Y    NaN   
2029          1.0        312.0          TA          TA            Y    NaN   
2203          2.0        560.0          TA          TA            Y    NaN   
2306          2.0        595.0          TA          TA            Y    NaN   
2552          2.0        631.0          TA          TA            Y    NaN   

     Date Sold  SalePrice  
869    11-2008   163000.0  
1019   03-2006   369900.0  
1867   05-2008   135000.0  
2029   06-2010   172000.0  
2203   07-2006   188000.0  
2306   06-2008   183000.0  
2552   08-2007   204000.0  

[7 rows x 37 columns]

We don't have any repeated rows in the above data. This is because our check for .duplicated defaults to keep = 'first" which means the first occurrence is kept as the original and True will only apply for every row after the first occurrence. If we wanted to keep the last row as the original and all else as True duplicates, we could use keep = False. If we do not want to designate an original row and want all identical rows included as duplicates, we can add keep=False to .duplicated().

In [ ]:
# Include the first row along with all dupicates
duplicated_rows = df.duplicated(keep=False)
duplicated_rows.sum()

14

In [ ]:
# Wwe can sort by some column to hhave them next to each other
df[duplicated_rows].sort_values("PID")

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
290   526351010        RL            81     14267   Pave   NaN    AllPub   
2029  526351010        RL            81     14267   Pave   NaN    AllPub   
1100  528174020        RL            34      3901   Pave   NaN    AllPub   
2552  528174020        RL            34      3901   Pave   NaN    AllPub   
524   535153150        RL            76      9120   Pave   NaN    AllPub   
869   535153150        RL            76      9120   Pave   NaN    AllPub   
1943  907262070        RL            72      7226   Pave   NaN    AllPub   
2306  907262070        RL            72      7226   Pave   NaN    AllPub   
1577  908103280        RL            65      6500   Pave   NaN    AllPub   
1867  908103280        RL            65      6500   Pave   NaN    AllPub   
950   921205030        RL            88     11443   Pave   NaN    AllPub   
1019  921205030        RL            88     11443   Pave   NaN    AllPub   
314   923230040        RL            63      9297   Pave   NaN    AllPub   
2203  923230040        RL            63      9297   Pave   NaN    AllPub   

     Neighborhood Bldg Type House Style  ...  Garage Type  Garage Yr Blt  \
290         NAmes      1Fam      1Story  ...       Attchd         1958.0   
2029        NAmes      1Fam      1Story  ...       Attchd         1958.0   
1100      NridgHt     Twnhs      1Story  ...       Attchd         2005.0   
2552      NridgHt     Twnhs      1Story  ...       Attchd         2005.0   
524         NAmes      1Fam      1Story  ...       Attchd         1958.0   
869         NAmes      1Fam      1Story  ...       Attchd         1958.0   
1943      CollgCr      1Fam      2Story  ...       Attchd         2003.0   
2306      CollgCr      1Fam      2Story  ...       Attchd         2003.0   
1577      Edwards      1Fam      1Story  ...       Detchd         1991.0   
1867      Edwards      1Fam      1Story  ...       Detchd         1991.0   
950        Timber      1Fam      1Story  ...       Attchd         2005.0   
1019       Timber      1Fam      1Story  ...       Attchd         2005.0   
314       Mitchel    Duplex      1Story  ...       Detchd         1976.0   
2203      Mitchel    Duplex      1Story  ...       Detchd         1976.0   

      Garage Cars  Garage Area Garage Qual Garage Cond  Paved Drive  Fence  \
290           1.0        312.0          TA          TA            Y    NaN   
2029          1.0        312.0          TA          TA            Y    NaN   
1100          2.0        631.0          TA          TA            Y    NaN   
2552          2.0        631.0          TA          TA            Y    NaN   
524           2.0        433.0          TA          TA            Y    NaN   
869           2.0        433.0          TA          TA            Y    NaN   
1943          2.0        595.0          TA          TA            Y    NaN   
2306          2.0        595.0          TA          TA            Y    NaN   
1577          2.0        480.0          TA          TA            Y    NaN   
1867          2.0        480.0          TA          TA            Y    NaN   
950           3.0        880.0          TA          TA            Y    NaN   
1019          3.0        880.0          TA          TA            Y    NaN   
314           2.0        560.0          TA          TA            Y    NaN   
2203          2.0        560.0          TA          TA            Y    NaN   

     Date Sold  SalePrice  
290    06-2010   172000.0  
2029   06-2010   172000.0  
1100   08-2007   204000.0  
2552   08-2007   204000.0  
524    11-2008   163000.0  
869    11-2008   163000.0  
1943   06-2008   183000.0  
2306   06-2008   183000.0  
1577   05-2008   135000.0  
1867   05-2008   135000.0  
950    03-2006   369900.0  
1019   03-2006   369900.0  
314    07-2006   188000.0  
2203   07-2006   188000.0  

[14 rows x 37 columns]

**Dropping Duplicate Rows**  

Once we identify duplicate rows, we should remove them. After dropping, we again check for duplicates to verify that we now have 0 duplicates.

In [ ]:
# Remove duplicates
df = df.drop_duplicates()

# We can also alter the dataframe directly
# df.drop_duplicates(inplace=True)

df.duplicated().sum()

0

#### Unique Values
Sometimes there may be rows that are essentially duplicate entries but with some small discrepancy. A single non-duplicated value in 1 column can cause .duplicated() to overlook the duplication.

One consideration for identifying such duplicates is if there is an identifying feature in the dataset that should be unique for each entry. In our case, the PID should not be repeated as each house should have its own value.

We can check to see how many unique values are included in each column using .nunique()

In [ ]:
# Check for how many unique values are in each column
df.nunique()

PID                 2930
MS Zoning              7
Lot Frontage         129
Lot Area            1960
Street                 2
Alley                  2
Utilities              3
Neighborhood          28
Bldg Type              5
House Style            8
Overall Qual          10
Overall Cond           9
Year Built           118
Year Remodeled        61
Exter Qual             4
Exter Cond             5
Bsmt Unf Sqft       1137
Total Bsmnt Sqft    1058
Central Air            4
Living Area Sqft    1292
Bsmt Full Bath         4
Bsmt Half Bath         3
Full Bath              5
Half Bath              4
Bedroom                8
Kitchen                4
Total Rooms           14
Garage Type            6
Garage Yr Blt        103
Garage Cars            6
Garage Area          603
Garage Qual            5
Garage Cond            5
Paved Drive            3
Fence                  4
Date Sold             55
SalePrice           1033
dtype: int64

In [ ]:
# We can also see the fraction of the data that is unique
df.nunique() / len(df)

# OR Percentage
df.nunique() / len(df) * 100

PID                 99.254743
MS Zoning            0.237127
Lot Frontage         4.369919
Lot Area            66.395664
Street               0.067751
Alley                0.067751
Utilities            0.101626
Neighborhood         0.948509
Bldg Type            0.169377
House Style          0.271003
Overall Qual         0.338753
Overall Cond         0.304878
Year Built           3.997290
Year Remodeled       2.066396
Exter Qual           0.135501
Exter Cond           0.169377
Bsmt Unf Sqft       38.516260
Total Bsmnt Sqft    35.840108
Central Air          0.135501
Living Area Sqft    43.766938
Bsmt Full Bath       0.135501
Bsmt Half Bath       0.101626
Full Bath            0.169377
Half Bath            0.135501
Bedroom              0.271003
Kitchen              0.135501
Total Rooms          0.474255
Garage Type          0.203252
Garage Yr Blt        3.489160
Garage Cars          0.203252
Garage Area         20.426829
Garage Qual          0.169377
Garage Cond          0.169377
Paved Driv

**Duplicates in a subset**  

We can apply .duplicated() to a subset of the data to check for duplicates in a particular column. In our case, we want to find rows that have duplicate PID values.

In [ ]:
df.duplicated(subset=['PID','MS Zoning']).sum()

22

In [ ]:
# How many rows are duplicates (including the first occurence)
duplicated_pids = df.duplicated(subset=['PID'], keep=False)
duplicated_pids.sum()

44

In [ ]:
# Sorting for better display
# Visually checking the duplicate rows
df[duplicated_pids].sort_values("PID").head()

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
2654  526355080        RL            75     13860   Pave   NaN    AllPub   
1650  526355080        RL            75     13860   Pave   NaN    AllPub   
135   527110020        RL            -1      8530   Pave   NaN    AllPub   
2469  527110020        RL            -1      8530   Pave   NaN    AllPub   
626   527326040        RL            85     11900   Pave   NaN    AllPub   

     Neighborhood Bldg Type House Style  ...  Garage Type  Garage Yr Blt  \
2654        NAmes      1Fam        SLvl  ...       Attchd         1972.0   
1650        NAmes      1Fam        SLvl  ...       Attchd         1972.0   
135       Gilbert      1Fam        SLvl  ...      BuiltIn         1995.0   
2469      Gilbert      1Fam        SLvl  ...      BuiltIn         1995.0   
626        NWAmes      1Fam      1Story  ...       Attchd         1977.0   

      Garage Cars  Garage Area Garage Qual Garage Cond  Paved Drive  Fence  \
2654          2.0        538.0          TA          TA            Y  MnPrv   
1650          2.0        538.0          TA          TA            Y  MnPrv   
135           2.0        400.0          TA          TA            Y    NaN   
2469          2.0        400.0          TA          TA            Y    NaN   
626           2.0        544.0          TA          TA            Y    NaN   

     Date Sold  SalePrice  
2654   07-2009        NaN  
1650   07-2009   345000.0  
135    05-2009   168500.0  
2469   05-2009        NaN  
626    04-2009    82500.0  

[5 rows x 37 columns]

The entries at index 2654 and 1650 have the same PID number. Scrolling all the way to the far right of the data frame, we see that the only difference is that one entry has a value for "SalePrice" while the other has a NaN for Sale Price.  
If you continue to inspect the rows with duplicate PIDs, you will find a similar result for all 22 cases.

# Identify Missing Values  
In a spreadsheet, a null value is just an empty cell. In NumPy and Pandas, null values will show up as "NaN," which stands for "Not a Number," but they are not strings either. This just means no data is included. NaNs are a subtype of float, and any integer columns with NaNs will be recast as floats.  
 The image below shows a null value (NaN) in a dataframe.  

 <img src="https://assets.codingdojo.com/boomyeah2015/codingdojo/curriculum/content/chapter/1680019526__Capture.PNG" width="220px" height="300px">

In [ ]:
# Finding if there are null values with isna() method
df.isna()

PID  MS Zoning  Lot Frontage  Lot Area  Street  Alley  Utilities  \
0     False      False         False     False   False   True      False   
1     False      False         False     False   False   True      False   
2     False      False         False     False   False   True      False   
3     False      False         False     False   False   True      False   
4     False      False         False     False   False  False      False   
...     ...        ...           ...       ...     ...    ...        ...   
2954  False      False         False     False   False  False      False   
2955  False      False         False     False   False   True      False   
2956  False      False         False     False   False   True      False   
2957  False      False         False     False   False   True      False   
2958  False      False         False     False   False   True      False   

      Neighborhood  Bldg Type  House Style  ...  Garage Type  Garage Yr Blt  \
0            False      False        False  ...        False          False   
1            False      False        False  ...        False          False   
2            False      False        False  ...        False          False   
3            False      False        False  ...        False          False   
4            False      False        False  ...        False          False   
...            ...        ...          ...  ...          ...            ...   
2954         False      False        False  ...        False          False   
2955         False      False        False  ...        False          False   
2956         False      False        False  ...        False          False   
2957         False      False        False  ...        False          False   
2958         False      False        False  ...        False          False   

      Garage Cars  Garage Area  Garage Qual  Garage Cond  Paved Drive  Fence  \
0           False        False        False        False        False  False   
1           False        False        False        False        False   True   
2           False        False        False        False        False   True   
3           False        False        False        False        False   True   
4           False        False        False        False        False   True   
...           ...          ...          ...          ...          ...    ...   
2954        False        False        False        False        False   True   
2955        False        False        False        False        False   True   
2956        False        False        False        False        False   True   
2957        False        False        False        False        False  False   
2958        False        False        False        False        False   True   

      Date Sold  SalePrice  
0         False      False  
1         False      False  
2         False      False  
3         False      False  
4         False      False  
...         ...        ...  
2954      False      False  
2955      False      False  
2956      False      False  
2957      False      False  
2958      False      False  

[2952 rows x 37 columns]

In [ ]:
df.isnull().sum()

PID                    0
MS Zoning              0
Lot Frontage           0
Lot Area               0
Street                 0
Alley               2751
Utilities              0
Neighborhood           0
Bldg Type              0
House Style            0
Overall Qual           0
Overall Cond           0
Year Built             0
Year Remodeled         0
Exter Qual             0
Exter Cond             0
Bsmt Unf Sqft          1
Total Bsmnt Sqft       1
Central Air            0
Living Area Sqft       0
Bsmt Full Bath         2
Bsmt Half Bath         2
Full Bath              0
Half Bath              0
Bedroom                0
Kitchen                0
Total Rooms            0
Garage Type          157
Garage Yr Blt        159
Garage Cars            1
Garage Area            1
Garage Qual          159
Garage Cond          159
Paved Drive            0
Fence               2378
Date Sold              0
SalePrice             22
dtype: int64

In [ ]:
df.info()

In [ ]:
# We can aggregate these to find how many values are missing
null_sums = df.isna().sum()
null_sums

PID                    0
MS Zoning              0
Lot Frontage           0
Lot Area               0
Street                 0
Alley               2751
Utilities              0
Neighborhood           0
Bldg Type              0
House Style            0
Overall Qual           0
Overall Cond           0
Year Built             0
Year Remodeled         0
Exter Qual             0
Exter Cond             0
Bsmt Unf Sqft          1
Total Bsmnt Sqft       1
Central Air            0
Living Area Sqft       0
Bsmt Full Bath         2
Bsmt Half Bath         2
Full Bath              0
Half Bath              0
Bedroom                0
Kitchen                0
Total Rooms            0
Garage Type          157
Garage Yr Blt        159
Garage Cars            1
Garage Area            1
Garage Qual          159
Garage Cond          159
Paved Drive            0
Fence               2378
Date Sold              0
SalePrice             22
dtype: int64

In [ ]:
# Get the percentage of null values per column
null_percent = null_sums/len(df) * 100
null_percent

PID                  0.000000
MS Zoning            0.000000
Lot Frontage         0.000000
Lot Area             0.000000
Street               0.000000
Alley               93.191057
Utilities            0.000000
Neighborhood         0.000000
Bldg Type            0.000000
House Style          0.000000
Overall Qual         0.000000
Overall Cond         0.000000
Year Built           0.000000
Year Remodeled       0.000000
Exter Qual           0.000000
Exter Cond           0.000000
Bsmt Unf Sqft        0.033875
Total Bsmnt Sqft     0.033875
Central Air          0.000000
Living Area Sqft     0.000000
Bsmt Full Bath       0.067751
Bsmt Half Bath       0.067751
Full Bath            0.000000
Half Bath            0.000000
Bedroom              0.000000
Kitchen              0.000000
Total Rooms          0.000000
Garage Type          5.318428
Garage Yr Blt        5.386179
Garage Cars          0.033875
Garage Area          0.033875
Garage Qual          5.386179
Garage Cond          5.386179
Paved Driv

In [ ]:
# Filter to see only those with null values
null_percent[null_percent>0]

Alley               93.191057
Bsmt Unf Sqft        0.033875
Total Bsmnt Sqft     0.033875
Bsmt Full Bath       0.067751
Bsmt Half Bath       0.067751
Garage Type          5.318428
Garage Yr Blt        5.386179
Garage Cars          0.033875
Garage Area          0.033875
Garage Qual          5.386179
Garage Cond          5.386179
Fence               80.555556
SalePrice            0.745257
dtype: float64

### Dropping Na Values

Now we will drop any of the rows that have NaN in the SalePrice column.

In [ ]:
# Confirming  no more duplicated PIDs
df.duplicated(subset=['PID'], keep=False).sum()

44

In [ ]:
# Revisiting the duplicate rows with null values from duplicates lesson
df[duplicated_pids].sort_values("PID")

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
2654  526355080        RL            75     13860   Pave   NaN    AllPub   
1650  526355080        RL            75     13860   Pave   NaN    AllPub   
135   527110020        RL            -1      8530   Pave   NaN    AllPub   
2469  527110020        RL            -1      8530   Pave   NaN    AllPub   
626   527326040        RL            85     11900   Pave   NaN    AllPub   
625   527326040        RL            85     11900   Pave   NaN    AllPub   
2341  528178070        RL           130     16900   Pave   NaN    AllPub   
929   528178070        RL           130     16900   Pave   NaN    AllPub   
2599  528429100        RL            49     15218   Pave   NaN    AllPub   
324   528429100        RL            49     15218   Pave   NaN    AllPub   
1660  528477070        FV            84     10603   Pave   NaN    AllPub   
1946  528477070        FV            84     10603   Pave   NaN    AllPub   
2867  531376070        RL            81      9260   Pave   NaN    AllPub   
1266  531376070        RL            81      9260   Pave   NaN    AllPub   
2219  533206060        FV            80      3523   Pave  Pave    AllPub   
1153  533206060        FV            80      3523   Pave  Pave    AllPub   
747   534104090        FV            -1      7050   Pave   NaN    AllPub   
2192  534104090        FV            -1      7050   Pave   NaN    AllPub   
801   535303050        RL            75     11310   Pave   NaN    AllPub   
185   535303050        RL            75     11310   Pave   NaN    AllPub   
692   535350040        RL            74      5868   Pave   NaN    AllPub   
602   535350040        RL            74      5868   Pave   NaN    AllPub   
938   902128160        RM            60      7200   Pave   NaN    AllPub   
2916  902128160        RM            60      7200   Pave   NaN    AllPub   
581   902202090        RM            75      9000   Pave   NaN    AllPub   
354   902202090        RM            75      9000   Pave   NaN    AllPub   
286   902300040        RM            53      9863   Pave  Pave    AllPub   
1303  902300040        RM            53      9863   Pave  Pave    AllPub   
51    903234160        RM            60      6180   Pave   NaN    AllPub   
1196  903234160        RM            60      6180   Pave   NaN    AllPub   
71    905476120        RL            50      8405   Pave  Grvl    AllPub   
2795  905476120        RL            50      8405   Pave  Grvl    AllPub   
2333  906380170        RL            64      7406   Pave   NaN    AllPub   
2706  906380170        RL            64      7406   Pave   NaN    AllPub   
2171  907262020        RL            65      8158   Pave   NaN    AllPub   
634   907262020        RL            65      8158   Pave   NaN    AllPub   
1093  907265100        RL            96     11690   Pave   NaN    AllPub   
1846  907265100        RL            96     11690   Pave   NaN    AllPub   
2460  909250080        RL            43      5707   Pave   NaN    AllPub   
88    909250080        RL            43      5707   Pave   NaN    AllPub   
1773  909251090        RL            79     11526   Pave   NaN    AllPub   
1474  909251090        RL            79     11526   Pave   NaN    AllPub   
1913  923228250        RM            21      2001   Pave   NaN    AllPub   
308   923228250        RM            21      2001   Pave   NaN    AllPub   

     Neighborhood Bldg Type House Style  ...  Garage Type  Garage Yr Blt  \
2654        NAmes      1Fam        SLvl  ...       Attchd         1972.0   
1650        NAmes      1Fam        SLvl  ...       Attchd         1972.0   
135       Gilbert      1Fam        SLvl  ...      BuiltIn         1995.0   
2469      Gilbert      1Fam        SLvl  ...      BuiltIn         1995.0   
626        NWAmes      1Fam      1Story  ...       Attchd         1977.0   
625        NWAmes      1Fam      1Story  ...       Attchd         1977.0   
2341      NridgHt      1Fam      2Story  ...       Attchd  

In [ ]:
# Dropping Null values from SalePrice
df = df.dropna(subset=["SalePrice"])

In [ ]:
# Confirming no more null sale prices
df['SalePrice'].isna().sum()

0

### Handling Null Values

We apply different tactics for addressing null values, depending on if the column is categorical or numeric.

We also use different tools/tactics when dealing with null values for Data Understanding/inspection vs. when we prepare data for a machine learning model.

The approaches demonstrated in this lesson should be used for data exploration only. We will return to this point at the beginning of the Intro to Machine Learning course.

Pandas has convenient methods for addressing null values. Throughout this lesson, we will demonstrate each of these methods:

1. Drop rows/columns with null values.
2. Fill in with a placeholder value.
3. Impute with a central value


## Dropping Rows
**When?**

If only a small percentage of rows are missing data you may want to just drop them.  There is no hard and fast rule, but one rule of thumb is if fewer than 2% of your rows are missing data, it might be a good idea to just drop them.

**Pros to dropping rows**:  
1. You do not introduce any errors into your data.  Any time you fill missing data with values that you are not certain are correct, you are introducing errors into your data.
2. The remaining rows keep all of their information.  If you drop columns, every row loses some information that might be potentially useful.  
**Cons to dropping rows**:  
1. You lose data when you drop rows, and more GOOD information is always better than less in data science.  We want to take out the trash when we can, but we don’t want to lose too much information!

Ideally, we want to keep as much of our data as possible, but that isn't always possible. One major exception is rows that have null values for the target column that we are trying to explain/predict.

When we previously checked for duplicates, we found rows with duplicate PIDs and that one of each pair of duplicate rows was missing the Sale Price. This is the code from the end of the lesson where we first addressed duplicate rows.

# Handling Null Values Extended

To further demonstrate how to handle missing values, we can use the [titanic](https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv) dataset.

In [ ]:
# Load the data
fileurl = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(fileurl)
# Note: you can use the link to load this data directly

In [ ]:
# Basic exploration
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
# What else can we explore?

Check Null values

In [ ]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [ ]:
# Lets drop null values
df_no_null = df.dropna()

How much data have we lost?

In [ ]:
print(f'The orginal dataframe had {df.shape[0]} rows whle the new dataframe has {df_no_null.shape[0]} rows')

The orginal dataframe had 891 rows whle the new dataframe has 183 rows


### Dropping columns
Sometimes you will run across columns that are missing a significant amount, or even most of their data.  Dropping the rows would cause you to lose too much information, and you just don’t have enough good data in that column to make an educated guess about how to fill the missing values.  In this case, you may want to just drop the column.  Sometimes a column just doesn’t have enough data in it to be useful.



**When should I drop a column?**  
It’s important to know what a column of data represents and how important it is to the thing you are analyzing or the problem you are trying to solve.  If it’s a column of low relevance and missing a lot of data, it might be worth just dropping it.  It’s also important to examine how much data is present in the column.  Even if the column could be important, is there enough information there to make it worth trying to save?

Once again, there is no one rule that always gives the right answer.  Dropping a column takes careful consideration of many factors. However, if you don’t know how important the column is and don’t have a subject matter expert, you could use 50% as a general rule.  If a column is missing more than 50% of the data, you might want to consider dropping it.  However, this number is very controversial in the data science field and you will find many other opinions in your research on the topic.  
**Pros to Dropping Columns**
1. You are not introducing errors into your data
2. You are not losing any rows from your dataset
3. You might be removing un-helpful columns, which can help in predictive modeling later.


**Cons to Dropping Column**
1. You can no longer do analysis of that feature.

2. You may be losing information from your dataset that is important or useful.



In [ ]:
# We can drop specific column/columns entirely
df_droppped_column = df.drop(columns='Cabin')
df_droppped_column.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
dtype: int64

In [ ]:
# We can drop all columns with null values
df_droppped_columns = df.dropna(axis=1)
df_droppped_columns.isna().sum()

# df.dropna(axis=1, inplace=True)

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
SibSp          0
Parch          0
Ticket         0
Fare           0
dtype: int64

In [ ]:
# We can specify a condition for dropping based on the percentage of missing values
df.dropna(axis=1, thresh=.25)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### Fill with placeholder values

Sometimes the best strategy is to fill the missing values with a new value.  If it is a categorical column, you might just fill the missing cells with a new category e.g. ‘missing’.  

**When should I fill missing values with a new category?**

You might consider this if there seems to be a pattern to the missing values, but that pattern doesn’t suggest one of the other values in that column.  For instance, if the values in the ‘Gender’ column were ‘Male’, ‘Female’ and some were missing.  You might think that people filling out the survey didn’t feel comfortable choosing either ‘Male’ or ‘Female’ for their gender.  You might fill those missing values with ‘other’ or ‘missing’ or ‘declined to answer’.  This is another time when subject matter knowledge can help you make good choices.

**Pros to Filling Data with a New Category**  
1. You can highlight patterns of missing data for analysis.  Often patterns of missing data offer important insights!
2. You may give a machine learning model clues about the data if there is, in fact, a pattern to the missing data.  Maybe those missing data are predictive of something?  
**Cons to Filling Data with a New Category**  
1. You may be missing a chance to correctly fill some missing data using imputation

In [ ]:
# Consider our dataframe  again
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

We can see that we have values mssing for the Cabin column. Let us explore what values are present.

In [ ]:
# Get value counts for Cabin
df['Cabin'].value_counts()

B96 B98        4
G6             4
C23 C25 C27    4
C22 C26        3
F33            3
              ..
E34            1
C7             1
C54            1
E36            1
C148           1
Name: Cabin, Length: 147, dtype: int64

This is a categorical column for which we can fill the missing data with a placeholder

In [ ]:
# Create new column where missing values are filled
df['Cabin_filled'] = df['Cabin'].fillna('MISSING')

In [ ]:
# Check null values
df.isna().sum()

PassengerId       0
Survived          0
Pclass            0
Name              0
Sex               0
Age             177
SibSp             0
Parch             0
Ticket            0
Fare              0
Cabin           687
Embarked          2
Cabin_filled      0
dtype: int64

In [ ]:
# Change pandas options to show multiple rows
pd.set_option('display.max_rows',200)

# Get the value_counts
df[["Cabin_filled"]].value_counts()

Cabin_filled   
MISSING            687
G6                   4
C23 C25 C27          4
B96 B98              4
E101                 3
D                    3
C22 C26              3
F33                  3
F2                   3
C65                  2
C124                 2
C123                 2
B77                  2
C92                  2
E121                 2
C126                 2
E24                  2
E25                  2
B58 B60              2
C125                 2
D36                  2
C2                   2
B51 B53 B55          2
D35                  2
D33                  2
D26                  2
C52                  2
D20                  2
D17                  2
C68                  2
C78                  2
C83                  2
B57 B59 B63 B66      2
C93                  2
B18                  2
E67                  2
B35                  2
B28                  2
B22                  2
F G73                2
B5                   2
E44                  2
F4                

Filling numeric values is a little different,  overall we need to identify the range of the data and fill with a non-existent but consistent value.  
For example we can set the missing age values to -1 since we do not expect any ages to have this value

In [ ]:
# Check descriptive statistics for age
df['Age'].describe().round(2)

count    714.00
mean      29.70
std       14.53
min        0.42
25%       20.12
50%       28.00
75%       38.00
max       80.00
Name: Age, dtype: float64

In [ ]:
df['Age_filled'] = df['Age'].fillna(-1)

# Check for null values
df.isna().sum()

PassengerId       0
Survived          0
Pclass            0
Name              0
Sex               0
Age             177
SibSp             0
Parch             0
Ticket            0
Fare              0
Cabin           687
Embarked          2
Cabin_filled      0
Age_filled        0
dtype: int64

# Impute Missing Values
Imputation means guessing the missing value. Usually imputation uses existing information to approximate what is missing.  

**Pros to Imputing Values**
1. You do not lose any data

**Cons to Imputing Values**
1. You are introducing errors into your data.

Adding non-exixtent values potentially means introducing errors to the data. Although there are several imputation strategies, for purposes of this module we can use the following guidelines.

1. Examine the data carefully to look for patterns in the missing data.  Are there correlations between the missing values in one column with values in other columns?
2. Consult a subject matter expert or use your own subject knowledge to understand the data and leverage that understanding to make educated guesses about how to fill the missing data.
3. Look for patterns in the data to suggest what data might belong in the missing cells.

For most cases though we can use some of these simple strategies:

### Imputing numerical data

Usually when dealing with numerical data, we consider the central value that best defines the data. We have previously covered mean, mode and median as central measures and these can be used.  
**NOTE: Use the measure that best describes the data. e.g. when you have very many outliers, median might be the best measure over mean**


In [ ]:
# Example with age - mean
mean_age = df['Age'].mean()
df['Age_mean_imputed'] = df['Age'].fillna(mean_age)

# Check for null values
df.isna().sum()

PassengerId           0
Survived              0
Pclass                0
Name                  0
Sex                   0
Age                 177
SibSp                 0
Parch                 0
Ticket                0
Fare                  0
Cabin               687
Embarked              2
Cabin_filled          0
Age_filled            0
Age_mean_imputed      0
dtype: int64

In [ ]:
# Example with age- median
median_age = df['Age'].median()
df['Age_median_imputed'] = df['Age'].fillna(median_age)

# Check for null values
df.isna().sum()

PassengerId             0
Survived                0
Pclass                  0
Name                    0
Sex                     0
Age                   177
SibSp                   0
Parch                   0
Ticket                  0
Fare                    0
Cabin                 687
Embarked                2
Cabin_filled            0
Age_filled              0
Age_mean_imputed        0
Age_median_imputed      0
dtype: int64

In [ ]:
# Example with age- mode
mode_age = df['Age'].mode()
df['Age_mode_imputed'] = df['Age'].fillna(mode_age[0])

# Check for null values
df.isna().sum()

PassengerId             0
Survived                0
Pclass                  0
Name                    0
Sex                     0
Age                   177
SibSp                   0
Parch                   0
Ticket                  0
Fare                    0
Cabin                 687
Embarked                2
Cabin_filled            0
Age_filled              0
Age_mean_imputed        0
Age_median_imputed      0
Age_mode_imputed        0
dtype: int64

In [ ]:
df.describe().round(3)

PassengerId  Survived   Pclass      Age    SibSp    Parch     Fare  \
count      891.000   891.000  891.000  714.000  891.000  891.000  891.000   
mean       446.000     0.384    2.309   29.699    0.523    0.382   32.204   
std        257.354     0.487    0.836   14.526    1.103    0.806   49.693   
min          1.000     0.000    1.000    0.420    0.000    0.000    0.000   
25%        223.500     0.000    2.000   20.125    0.000    0.000    7.910   
50%        446.000     0.000    3.000   28.000    0.000    0.000   14.454   
75%        668.500     1.000    3.000   38.000    1.000    0.000   31.000   
max        891.000     1.000    3.000   80.000    8.000    6.000  512.329   

       Age_filled  Age_mean_imputed  Age_median_imputed  Age_mode_imputed  
count     891.000           891.000             891.000           891.000  
mean       23.601            29.699              29.362            28.567  
std        17.867            13.002              13.020            13.200  
min        -1.000             0.420               0.420             0.420  
25%         6.000            22.000              22.000            22.000  
50%        24.000            29.699              28.000            24.000  
75%        35.000            35.000              35.000            35.000  
max        80.000            80.000              80.000            80.000

## Imputing Categorical Data

When dealing with categorical data, the most common measure of central tendency is mode

In [ ]:
# Imputing the Cabin with mode

mode_cabin = df['Cabin'].mode()
df['Cabin_mode_imputed'] = df['Cabin'].fillna(mode_cabin[0])

# Check null
df.isna().sum()

PassengerId             0
Survived                0
Pclass                  0
Name                    0
Sex                     0
Age                   177
SibSp                   0
Parch                   0
Ticket                  0
Fare                    0
Cabin                 687
Embarked                2
Cabin_filled            0
Age_filled              0
Age_mean_imputed        0
Age_median_imputed      0
Age_mode_imputed        0
Cabin_mode_imputed      0
dtype: int64

**NOTE: We have handled the data this way for descriptive analytical purposes only**

## Reading(Visualizing Null Values with missingno)
Sometimes missing values are just random omissions, but in some cases, there is a pattern related to missing values. To help us identify any patterns, we can create a visual of missing data.

missingno is a visualization package designed to show null values (missing numbers).

We can create a visual representation of our dataframe, where null values are represented graphically. While there are several plotting functions available, we will use the matrix plot.

We must first import the new library, and then we can run the matrix function on the dataframe:

In [ ]:
import missingno as msno
msno.matrix(df);

In [ ]:
# save the filter
# null_garage_type = df['Garage Type'].isna()
# null_garage_type



In [ ]:
# # Use the filter
# df[null_garage_type]
# # df[~null_garage_type] # Inverse with ~

In [ ]:
# df.describe(percentiles=[0.25])
# df["Overall Qual"].describe()